Add `rust-docs-mcp doctor` command for system diagnostics

# Add `rust-docs-mcp doctor` command for system diagnostics

## Overview
Implement a `doctor` subcommand that verifies the system environment and dependencies required for rust-docs-mcp to function properly. This will help users troubleshoot installation and runtime issues.

## Requirements
- Verify Rust toolchain installation (stable + nightly)
- Check Git availability and network connectivity
- Validate cache directory writability
- Test rustdoc JSON generation capability
- Exit with non-zero code if any checks fail
- Provide clear, actionable error messages

## Implementation Details

### Command Structure
```bash
rust-docs-mcp doctor [OPTIONS]
```

### Checks to Implement
1. **Rust Toolchain Verification**
   - Check `rustc --version` works
   - Verify nightly toolchain is installed: `rustup toolchain list | grep nightly`
   - Test rustdoc JSON generation: `rustdoc --version` and basic JSON output test

2. **Git Installation**
   - Verify `git --version` works
   - Test basic git operations (clone capability)

3. **Network Connectivity**
   - Test connection to crates.io API
   - Test connection to GitHub (for repo cloning)

4. **Cache Directory**
   - Check if cache directory exists or can be created
   - Test write permissions
   - Check available disk space (warn if < 1GB)

5. **Optional Dependencies**
   - Check for `codesign` on macOS (for binary signing)

### Output Format
```
🔍 rust-docs-mcp doctor

✅ Rust toolchain: rustc 1.75.0 (stable)
✅ Nightly toolchain: rustc 1.76.0-nightly
✅ Git: git version 2.39.0
✅ Network: crates.io reachable
✅ Cache directory: ~/.rust-docs-mcp/cache (writable, 15GB available)
❌ Rustdoc JSON: Failed to generate test documentation

[ERROR] Rustdoc JSON generation failed. Please ensure nightly toolchain is properly installed:
  rustup toolchain install nightly

Doctor found 1 issue. Please fix the above errors before using rust-docs-mcp.
```

### Error Exit Codes
- 0: All checks passed
- 1: One or more checks failed
- 2: Critical system dependency missing

## Files to Modify
- `src/main.rs`: Add `Doctor` command to `Commands` enum
- `src/main.rs`: Add `handle_doctor_command()` function
- `src/doctor.rs`: New module with diagnostic functions

## Estimated Effort
≤ 200 LoC - Single module with system check functions

## Acceptance Criteria
- [ ] `rust-docs-mcp doctor` command exists
- [ ] Verifies all critical dependencies
- [ ] Provides actionable error messages
- [ ] Exits with appropriate status codes
- [ ] Includes clear success/failure indicators
- [ ] Handles edge cases gracefully (no network, no permissions, etc.)

--- a/README.md
+++ b/README.md
@@ -147,8 +147,29 @@ cargo build --release
 rust-docs-mcp                   # Start MCP server
 rust-docs-mcp install           # Install to ~/.local/bin
 rust-docs-mcp install --force   # Force overwrite existing installation
+rust-docs-mcp doctor            # Verify system environment and dependencies
+rust-docs-mcp doctor --json     # Output diagnostic results in JSON format
+rust-docs-mcp update            # Update to latest version from GitHub
 rust-docs-mcp --help            # Show help
 ```
+
+### Troubleshooting
+
+If you encounter issues during installation or runtime, run the doctor command to diagnose common problems:
+
+```bash
+rust-docs-mcp doctor
+```
+
+The doctor command checks:
+- Rust toolchain availability (stable + nightly)
+- Git installation
+- Network connectivity to crates.io and GitHub
+- Cache directory permissions and disk space
+- Rustdoc JSON generation capability
+- Optional dependencies (e.g., codesign on macOS)
+
+For programmatic integration, use `--json` flag to get structured output.
 
 ### MCP Configuration
 

--- a/rust-docs-mcp/Cargo.toml
+++ b/rust-docs-mcp/Cargo.toml
@@ -52,5 +52,6 @@ tracing-subscriber = { version = "0.3", features = [
     "fmt",
 ] }
 tempfile = "3.8"
+fs2 = "0.4"
 
 [dev-dependencies]

--- a/rust-docs-mcp/src/cache/docgen.rs
+++ b/rust-docs-mcp/src/cache/docgen.rs
@@ -47,7 +47,7 @@ impl DocGenerator {
         tracing::info!("Generating documentation for {}-{}", name, version);
 
         // Run cargo rustdoc with JSON output using unified function
-        rustdoc::run_cargo_rustdoc_json(&source_path).await?;
+        rustdoc::run_cargo_rustdoc_json(&source_path, None).await?;
 
         // Find the generated JSON file in target/doc
         let doc_dir = source_path.join("target").join("doc");
@@ -115,7 +115,7 @@ impl DocGenerator {
         );
 
         // Run cargo rustdoc with JSON output for the specific package using unified function
-        rustdoc::run_cargo_rustdoc_json_for_package(&source_path, &package_name).await?;
+        rustdoc::run_cargo_rustdoc_json(&source_path, Some(&package_name)).await?;
 
         // Find the generated JSON file in target/doc
         let doc_dir = source_path.join("target").join("doc");

--- a/rust-docs-mcp/src/cache/docgen.rs
+++ b/rust-docs-mcp/src/cache/docgen.rs
@@ -5,12 +5,11 @@
 
 use crate::cache::storage::CacheStorage;
 use crate::cache::workspace::WorkspaceHandler;
+use crate::rustdoc;
 use anyhow::{Context, Result, bail};
 use std::path::{Path, PathBuf};
 use std::process::Command;
 
-/// The pinned nightly toolchain version compatible with rustdoc-types 0.53.0
-const REQUIRED_TOOLCHAIN: &str = "nightly-2025-06-23";
 
 /// Service for generating documentation from Rust crates
 #[derive(Debug, Clone)]
@@ -26,26 +25,7 @@ impl DocGenerator {
 
     /// Validate that the required toolchain is available
     async fn validate_toolchain(&self) -> Result<()> {
-        let output = Command::new("rustup")
-            .args(["toolchain", "list"])
-            .output()
-            .context("Failed to run rustup toolchain list")?;
-
-        if !output.status.success() {
-            bail!("Failed to check available toolchains");
-        }
-
-        let toolchains = String::from_utf8_lossy(&output.stdout);
-        if !toolchains.contains(REQUIRED_TOOLCHAIN) {
-            bail!(
-                "Required toolchain {} is not installed. Please run: rustup toolchain install {}",
-                REQUIRED_TOOLCHAIN,
-                REQUIRED_TOOLCHAIN
-            );
-        }
-
-        tracing::debug!("Validated toolchain {} is available", REQUIRED_TOOLCHAIN);
-        Ok(())
+        rustdoc::validate_toolchain().await
     }
 
     /// Generate JSON documentation for a crate
@@ -66,26 +46,8 @@ impl DocGenerator {
 
         tracing::info!("Generating documentation for {}-{}", name, version);
 
-        // Run cargo rustdoc with JSON output using pinned nightly toolchain
-        let output = Command::new("cargo")
-            .args([
-                &format!("+{}", REQUIRED_TOOLCHAIN),
-                "rustdoc",
-                "--all-features",
-                "--",
-                "--output-format",
-                "json",
-                "-Z",
-                "unstable-options",
-            ])
-            .current_dir(&source_path)
-            .output()
-            .context("Failed to run cargo rustdoc")?;
-
-        if !output.status.success() {
-            let stderr = String::from_utf8_lossy(&output.stderr);
-            bail!("Failed to generate documentation: {}", stderr);
-        }
+        // Run cargo rustdoc with JSON output using unified function
+        rustdoc::run_cargo_rustdoc_json(&source_path).await?;
 
         // Find the generated JSON file in target/doc
         let doc_dir = source_path.join("target").join("doc");
@@ -152,28 +114,8 @@ impl DocGenerator {
             version
         );
 
-        // Run cargo rustdoc with JSON output for the specific package using pinned nightly toolchain
-        let output = Command::new("cargo")
-            .args([
-                &format!("+{}", REQUIRED_TOOLCHAIN),
-                "rustdoc",
-                "-p",
-                &package_name,
-                "--all-features",
-                "--",
-                "--output-format",
-                "json",
-                "-Z",
-                "unstable-options",
-            ])
-            .current_dir(&source_path)
-            .output()
-            .context("Failed to run cargo rustdoc")?;
-
-        if !output.status.success() {
-            let stderr = String::from_utf8_lossy(&output.stderr);
-            bail!("Failed to generate documentation: {}", stderr);
-        }
+        // Run cargo rustdoc with JSON output for the specific package using unified function
+        rustdoc::run_cargo_rustdoc_json_for_package(&source_path, &package_name).await?;
 
         // Find the generated JSON file in target/doc
         let doc_dir = source_path.join("target").join("doc");

--- a/rust-docs-mcp/src/doctor.rs
+++ b/rust-docs-mcp/src/doctor.rs
@@ -1,10 +1,9 @@
-use anyhow::{anyhow, Result};
+use anyhow::Result;
 use std::process::Command;
-use std::path::Path;
 use std::fs;
 use reqwest;
 use dirs;
-use tempfile;
+use crate::rustdoc;
 
 pub struct DiagnosticResult {
     pub name: String,
@@ -126,72 +125,29 @@ async fn check_nightly_toolchain() -> DiagnosticResult {
 
 async fn check_rustdoc_json() -> DiagnosticResult {
     // First check if rustdoc is available
-    match Command::new("rustdoc").arg("--version").output() {
-        Ok(output) if output.status.success() => {
-            let version = String::from_utf8_lossy(&output.stdout).trim().to_string();
-            
-            // Try to create a simple test for JSON generation
-            let temp_dir = match tempfile::tempdir() {
-                Ok(dir) => dir,
-                Err(_) => return DiagnosticResult::new(
-                    "Rustdoc JSON".to_string(),
-                    false,
-                    "Failed to create temporary directory for testing".to_string(),
-                    false,
-                ),
-            };
-            
-            let test_file = temp_dir.path().join("lib.rs");
-            if let Err(_) = fs::write(&test_file, "//! Test crate\npub fn test() {}") {
-                return DiagnosticResult::new(
-                    "Rustdoc JSON".to_string(),
-                    false,
-                    "Failed to create test file".to_string(),
-                    false,
-                );
-            }
-            
-            // Try to generate JSON documentation
-            let test_file_str = match test_file.to_str() {
-                Some(path) => path,
-                None => return DiagnosticResult::new(
-                    "Rustdoc JSON".to_string(),
-                    false,
-                    "Test file path contains invalid UTF-8".to_string(),
-                    false,
-                ),
-            };
-            
-            match Command::new("rustdoc")
-                .args(&[
-                    "+nightly",
-                    "--output-format", "json",
-                    "--crate-name", "test",
-                    test_file_str,
-                ])
-                .output() {
-                Ok(json_output) if json_output.status.success() => {
+    match rustdoc::get_rustdoc_version().await {
+        Ok(version) => {
+            // Try to test JSON generation using the unified function
+            match rustdoc::test_rustdoc_json().await {
+                Ok(_) => {
                     DiagnosticResult::new(
                         "Rustdoc JSON".to_string(),
                         true,
-                        format!("{} with JSON support", version),
+                        format!("{} with JSON support (toolchain: {})", version, rustdoc::REQUIRED_TOOLCHAIN),
                         false,
                     )
                 }
-                _ => DiagnosticResult::new(
-                    "Rustdoc JSON".to_string(),
-                    false,
-                    "JSON generation failed - ensure nightly toolchain is installed".to_string(),
-                    false,
-                ),
+                Err(e) => {
+                    tracing::debug!("Rustdoc JSON test failed: {}", e);
+                    DiagnosticResult::new(
+                        "Rustdoc JSON".to_string(),
+                        false,
+                        format!("JSON generation failed: {}", e),
+                        false,
+                    )
+                }
             }
         }
-        Ok(_) => DiagnosticResult::new(
-            "Rustdoc JSON".to_string(),
-            false,
-            "rustdoc command failed".to_string(),
-            false,
-        ),
         Err(_) => DiagnosticResult::new(
             "Rustdoc JSON".to_string(),
             false,
@@ -228,38 +184,94 @@ async fn check_git_installation() -> DiagnosticResult {
 }
 
 async fn check_network_connectivity() -> DiagnosticResult {
-    let client = reqwest::Client::builder()
+    let client = match reqwest::Client::builder()
         .timeout(std::time::Duration::from_secs(10))
-        .build()
-        .unwrap();
-    
-    // Test crates.io API
-    match client.get("https://crates.io/api/v1/crates/serde").send().await {
-        Ok(response) if response.status().is_success() => {
-            // Also test GitHub connectivity
-            match client.get("https://api.github.com").send().await {
-                Ok(gh_response) if gh_response.status().is_success() => {
-                    DiagnosticResult::new(
-                        "Network".to_string(),
-                        true,
-                        "crates.io and GitHub reachable".to_string(),
-                        false,
-                    )
-                }
-                _ => DiagnosticResult::new(
-                    "Network".to_string(),
-                    false,
-                    "crates.io reachable but GitHub unreachable".to_string(),
-                    false,
-                ),
-            }
-        }
-        _ => DiagnosticResult::new(
+        .user_agent("rust-docs-mcp-doctor/1.0")
+        .build() {
+        Ok(client) => client,
+        Err(e) => return DiagnosticResult::new(
             "Network".to_string(),
             false,
-            "Unable to reach crates.io - check network connection".to_string(),
+            format!("Failed to create HTTP client: {}", e),
             false,
         ),
+    };
+    
+    // Test crates.io API
+    tracing::debug!("Testing crates.io connectivity...");
+    match client.get("https://crates.io/api/v1/crates/serde").send().await {
+        Ok(response) => {
+            let status = response.status();
+            tracing::debug!("crates.io response status: {}", status);
+            
+            if status.is_success() {
+                // Try to read a small portion of the response to ensure it's valid
+                match response.text().await {
+                    Ok(body) => {
+                        tracing::debug!("crates.io response body length: {}", body.len());
+                        
+                        // Also test GitHub connectivity
+                        tracing::debug!("Testing GitHub connectivity...");
+                        match client.get("https://api.github.com").send().await {
+                            Ok(gh_response) => {
+                                let gh_status = gh_response.status();
+                                tracing::debug!("GitHub response status: {}", gh_status);
+                                
+                                if gh_status.is_success() {
+                                    DiagnosticResult::new(
+                                        "Network".to_string(),
+                                        true,
+                                        format!("crates.io ({}) and GitHub ({}) reachable", status, gh_status),
+                                        false,
+                                    )
+                                } else {
+                                    DiagnosticResult::new(
+                                        "Network".to_string(),
+                                        false,
+                                        format!("crates.io reachable ({}) but GitHub unreachable ({})", status, gh_status),
+                                        false,
+                                    )
+                                }
+                            }
+                            Err(e) => {
+                                eprintln!("DEBUG: GitHub request error: {}", e);
+                                DiagnosticResult::new(
+                                    "Network".to_string(),
+                                    false,
+                                    format!("crates.io reachable ({}) but GitHub error: {}", status, e),
+                                    false,
+                                )
+                            }
+                        }
+                    }
+                    Err(e) => {
+                        eprintln!("DEBUG: Failed to read crates.io response body: {}", e);
+                        DiagnosticResult::new(
+                            "Network".to_string(),
+                            false,
+                            format!("crates.io responded ({}) but failed to read response: {}", status, e),
+                            false,
+                        )
+                    }
+                }
+            } else {
+                DiagnosticResult::new(
+                    "Network".to_string(),
+                    false,
+                    format!("crates.io returned error status: {}", status),
+                    false,
+                )
+            }
+        }
+        Err(e) => {
+            eprintln!("DEBUG: crates.io request error: {}", e);
+            DiagnosticResult::new(
+                "Network".to_string(),
+                false,
+                format!("Unable to reach crates.io: {}", e),
+                false,
+            )
+        }
     }
 }
 

--- a/rust-docs-mcp/src/doctor.rs
+++ b/rust-docs-mcp/src/doctor.rs
@@ -3,6 +3,7 @@ use std::process::Command;
 use std::fs;
 use reqwest;
 use dirs;
+use fs2;
 use crate::rustdoc;
 
 pub struct DiagnosticResult {
@@ -311,10 +312,30 @@ async fn check_cache_directory(cache_dir: Option<std::path::PathBuf>) -> Diagnos
             let _ = fs::remove_file(&test_file);
             
             // Check available disk space
-            match fs::metadata(&cache_path) {
-                Ok(_) => {
-                    // We can't easily get disk space in a cross-platform way without additional dependencies
-                    // For now, just confirm it's writable
+            match fs2::available_space(&cache_path) {
+                Ok(available_bytes) => {
+                    let available_formatted = format_bytes(available_bytes);
+                    // Warn if less than 1GB available
+                    if available_bytes < 1_073_741_824 {
+                        DiagnosticResult::new(
+                            "Cache directory".to_string(),
+                            false,
+                            format!("{} (writable, but only {} available - at least 1GB recommended)", 
+                                cache_path.display(), available_formatted),
+                            false,
+                        )
+                    } else {
+                        DiagnosticResult::new(
+                            "Cache directory".to_string(),
+                            true,
+                            format!("{} (writable, {} available)", cache_path.display(), available_formatted),
+                            false,
+                        )
+                    }
+                }
+                Err(e) => {
+                    // If disk space check fails, just report that it's writable
+                    tracing::debug!("Failed to check disk space: {}", e);
                     DiagnosticResult::new(
                         "Cache directory".to_string(),
                         true,
@@ -322,12 +343,6 @@ async fn check_cache_directory(cache_dir: Option<std::path::PathBuf>) -> Diagnos
                         false,
                     )
                 }
-                Err(_) => DiagnosticResult::new(
-                    "Cache directory".to_string(),
-                    true,
-                    format!("{} (writable)", cache_path.display()),
-                    false,
-                ),
             }
         }
         Err(e) => DiagnosticResult::new(
@@ -366,6 +381,23 @@ async fn check_optional_dependencies() -> DiagnosticResult {
         messages.join(", "),
         false,
     )
+}
+
+fn format_bytes(bytes: u64) -> String {
+    const UNITS: &[&str] = &["B", "KB", "MB", "GB", "TB"];
+    let mut size = bytes as f64;
+    let mut unit_index = 0;
+    
+    while size >= 1024.0 && unit_index < UNITS.len() - 1 {
+        size /= 1024.0;
+        unit_index += 1;
+    }
+    
+    if unit_index == 0 {
+        format!("{} {}", size as u64, UNITS[unit_index])
+    } else {
+        format!("{:.2} {}", size, UNITS[unit_index])
+    }
 }
 
 pub fn print_results(results: &[DiagnosticResult]) {
@@ -412,6 +444,9 @@ pub fn print_results(results: &[DiagnosticResult]) {
                     }
                     "Cache directory" => {
                         println!("\nCache directory issues detected. Check file permissions and disk space.");
+                        if result.message.contains("available") && result.message.contains("recommended") {
+                            println!("Consider freeing up disk space. At least 1GB is recommended for caching documentation.");
+                        }
                     }
                     _ => {}
                 }
@@ -545,5 +580,17 @@ mod tests {
         ];
         // This will print to stdout, but we're mainly testing it doesn't panic
         print_results(&results);
+    }
+    
+    #[test]
+    fn test_format_bytes() {
+        assert_eq!(format_bytes(0), "0 B");
+        assert_eq!(format_bytes(1023), "1023 B");
+        assert_eq!(format_bytes(1024), "1.00 KB");
+        assert_eq!(format_bytes(1536), "1.50 KB");
+        assert_eq!(format_bytes(1048576), "1.00 MB");
+        assert_eq!(format_bytes(1073741824), "1.00 GB");
+        assert_eq!(format_bytes(1099511627776), "1.00 TB");
+        assert_eq!(format_bytes(2147483648), "2.00 GB");
     }
 }

--- a/rust-docs-mcp/src/doctor.rs
+++ b/rust-docs-mcp/src/doctor.rs
@@ -4,8 +4,10 @@ use std::fs;
 use reqwest;
 use dirs;
 use fs2;
+use serde::Serialize;
 use crate::rustdoc;
 
+#[derive(Serialize)]
 pub struct DiagnosticResult {
     pub name: String,
     pub success: bool,
@@ -479,6 +481,22 @@ pub fn exit_code(results: &[DiagnosticResult]) -> i32 {
     } else {
         0 // All checks passed
     }
+}
+
+pub fn print_results_json(results: &[DiagnosticResult]) -> Result<()> {
+    let json_output = serde_json::json!({
+        "results": results,
+        "summary": {
+            "total_checks": results.len(),
+            "passed": results.iter().filter(|r| r.success).count(),
+            "failed": results.iter().filter(|r| !r.success).count(),
+            "critical_failures": results.iter().filter(|r| !r.success && r.critical).count(),
+        },
+        "exit_code": exit_code(results),
+    });
+    
+    println!("{}", serde_json::to_string_pretty(&json_output)?);
+    Ok(())
 }
 
 #[cfg(test)]

--- a/rust-docs-mcp/src/doctor.rs
+++ b/rust-docs-mcp/src/doctor.rs
@@ -499,6 +499,24 @@ pub fn print_results_json(results: &[DiagnosticResult]) -> Result<()> {
     Ok(())
 }
 
+/// Run diagnostics and print results with status message
+/// This is a convenience function used after install/update operations
+pub async fn run_and_print_diagnostics() -> Result<()> {
+    println!("\n🔍 Running system diagnostics...\n");
+    let results = run_diagnostics(None).await?;
+    print_results(&results);
+    
+    let exit_code = exit_code(&results);
+    if exit_code != 0 {
+        println!("\n⚠️  Some diagnostic checks failed. Please address the issues above.");
+        println!("You can run 'rust-docs-mcp doctor' anytime to check system status.");
+    } else {
+        println!("\n✅ All system checks passed!");
+    }
+    
+    Ok(())
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;

--- a/rust-docs-mcp/src/doctor.rs
+++ b/rust-docs-mcp/src/doctor.rs
@@ -4,6 +4,7 @@ use std::path::Path;
 use std::fs;
 use reqwest;
 use dirs;
+use tempfile;
 
 pub struct DiagnosticResult {
     pub name: String,
@@ -151,12 +152,22 @@ async fn check_rustdoc_json() -> DiagnosticResult {
             }
             
             // Try to generate JSON documentation
+            let test_file_str = match test_file.to_str() {
+                Some(path) => path,
+                None => return DiagnosticResult::new(
+                    "Rustdoc JSON".to_string(),
+                    false,
+                    "Test file path contains invalid UTF-8".to_string(),
+                    false,
+                ),
+            };
+            
             match Command::new("rustdoc")
                 .args(&[
                     "+nightly",
                     "--output-format", "json",
                     "--crate-name", "test",
-                    test_file.to_str().unwrap(),
+                    test_file_str,
                 ])
                 .output() {
                 Ok(json_output) if json_output.status.success() => {
@@ -420,5 +431,107 @@ pub fn exit_code(results: &[DiagnosticResult]) -> i32 {
         1 // One or more checks failed
     } else {
         0 // All checks passed
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_diagnostic_result_creation() {
+        let result = DiagnosticResult::new(
+            "Test".to_string(),
+            true,
+            "Test passed".to_string(),
+            false,
+        );
+        assert_eq!(result.name, "Test");
+        assert!(result.success);
+        assert_eq!(result.message, "Test passed");
+        assert!(!result.critical);
+    }
+
+    #[test]
+    fn test_exit_code_all_success() {
+        let results = vec![
+            DiagnosticResult::new("Test 1".to_string(), true, "Success".to_string(), false),
+            DiagnosticResult::new("Test 2".to_string(), true, "Success".to_string(), true),
+        ];
+        assert_eq!(exit_code(&results), 0);
+    }
+
+    #[test]
+    fn test_exit_code_non_critical_failure() {
+        let results = vec![
+            DiagnosticResult::new("Test 1".to_string(), true, "Success".to_string(), false),
+            DiagnosticResult::new("Test 2".to_string(), false, "Failed".to_string(), false),
+        ];
+        assert_eq!(exit_code(&results), 1);
+    }
+
+    #[test]
+    fn test_exit_code_critical_failure() {
+        let results = vec![
+            DiagnosticResult::new("Test 1".to_string(), false, "Failed".to_string(), true),
+            DiagnosticResult::new("Test 2".to_string(), false, "Failed".to_string(), false),
+        ];
+        assert_eq!(exit_code(&results), 2);
+    }
+
+    #[tokio::test]
+    async fn test_check_rust_toolchain() {
+        // This test will pass if rustc is installed
+        let result = check_rust_toolchain().await;
+        assert_eq!(result.name, "Rust toolchain");
+        // We can't guarantee the success state in all environments
+        // but we can verify it returns a valid DiagnosticResult
+        assert!(result.critical);
+    }
+
+    #[tokio::test]
+    async fn test_check_git_installation() {
+        // This test will pass if git is installed
+        let result = check_git_installation().await;
+        assert_eq!(result.name, "Git");
+        assert!(result.critical);
+    }
+
+    #[tokio::test]
+    async fn test_cache_directory_with_none() {
+        let result = check_cache_directory(None).await;
+        assert_eq!(result.name, "Cache directory");
+        // The success depends on whether the directory can be created
+        // but we can verify it returns a valid DiagnosticResult
+        assert!(!result.critical);
+    }
+
+    #[tokio::test]
+    async fn test_cache_directory_with_temp_dir() {
+        let temp_dir = tempfile::tempdir().unwrap();
+        let result = check_cache_directory(Some(temp_dir.path().to_path_buf())).await;
+        assert_eq!(result.name, "Cache directory");
+        assert!(result.success);
+        assert!(!result.critical);
+    }
+
+    #[tokio::test]
+    async fn test_optional_dependencies() {
+        let result = check_optional_dependencies().await;
+        assert_eq!(result.name, "Optional dependencies");
+        // Optional dependencies should always return success
+        assert!(result.success);
+        assert!(!result.critical);
+    }
+
+    #[test]
+    fn test_print_results_output() {
+        // This is a simple test to ensure print_results doesn't panic
+        let results = vec![
+            DiagnosticResult::new("Test 1".to_string(), true, "Success".to_string(), false),
+            DiagnosticResult::new("Test 2".to_string(), false, "Failed".to_string(), true),
+        ];
+        // This will print to stdout, but we're mainly testing it doesn't panic
+        print_results(&results);
     }
 }

--- a/rust-docs-mcp/src/doctor.rs
+++ b/rust-docs-mcp/src/doctor.rs
@@ -1,0 +1,424 @@
+use anyhow::{anyhow, Result};
+use std::process::Command;
+use std::path::Path;
+use std::fs;
+use reqwest;
+use dirs;
+
+pub struct DiagnosticResult {
+    pub name: String,
+    pub success: bool,
+    pub message: String,
+    pub critical: bool,
+}
+
+impl DiagnosticResult {
+    pub fn new(name: String, success: bool, message: String, critical: bool) -> Self {
+        Self {
+            name,
+            success,
+            message,
+            critical,
+        }
+    }
+}
+
+pub async fn run_diagnostics(cache_dir: Option<std::path::PathBuf>) -> Result<Vec<DiagnosticResult>> {
+    let mut results = Vec::new();
+    
+    // Check Rust toolchain
+    results.push(check_rust_toolchain().await);
+    
+    // Check nightly toolchain
+    results.push(check_nightly_toolchain().await);
+    
+    // Check rustdoc JSON capability
+    results.push(check_rustdoc_json().await);
+    
+    // Check Git installation
+    results.push(check_git_installation().await);
+    
+    // Check network connectivity
+    results.push(check_network_connectivity().await);
+    
+    // Check cache directory
+    results.push(check_cache_directory(cache_dir).await);
+    
+    // Check optional dependencies
+    results.push(check_optional_dependencies().await);
+    
+    Ok(results)
+}
+
+async fn check_rust_toolchain() -> DiagnosticResult {
+    match Command::new("rustc").arg("--version").output() {
+        Ok(output) if output.status.success() => {
+            let version = String::from_utf8_lossy(&output.stdout).trim().to_string();
+            DiagnosticResult::new(
+                "Rust toolchain".to_string(),
+                true,
+                version,
+                true,
+            )
+        }
+        Ok(_) => DiagnosticResult::new(
+            "Rust toolchain".to_string(),
+            false,
+            "rustc command failed".to_string(),
+            true,
+        ),
+        Err(_) => DiagnosticResult::new(
+            "Rust toolchain".to_string(),
+            false,
+            "rustc not found in PATH".to_string(),
+            true,
+        ),
+    }
+}
+
+async fn check_nightly_toolchain() -> DiagnosticResult {
+    match Command::new("rustup").args(&["toolchain", "list"]).output() {
+        Ok(output) if output.status.success() => {
+            let toolchains = String::from_utf8_lossy(&output.stdout);
+            if toolchains.contains("nightly") {
+                // Try to get nightly version
+                match Command::new("rustc").args(&["+nightly", "--version"]).output() {
+                    Ok(nightly_output) if nightly_output.status.success() => {
+                        let version = String::from_utf8_lossy(&nightly_output.stdout).trim().to_string();
+                        DiagnosticResult::new(
+                            "Nightly toolchain".to_string(),
+                            true,
+                            version,
+                            true,
+                        )
+                    }
+                    _ => DiagnosticResult::new(
+                        "Nightly toolchain".to_string(),
+                        false,
+                        "nightly toolchain installed but not functional".to_string(),
+                        true,
+                    ),
+                }
+            } else {
+                DiagnosticResult::new(
+                    "Nightly toolchain".to_string(),
+                    false,
+                    "nightly toolchain not installed".to_string(),
+                    true,
+                )
+            }
+        }
+        Ok(_) => DiagnosticResult::new(
+            "Nightly toolchain".to_string(),
+            false,
+            "rustup command failed".to_string(),
+            true,
+        ),
+        Err(_) => DiagnosticResult::new(
+            "Nightly toolchain".to_string(),
+            false,
+            "rustup not found in PATH".to_string(),
+            true,
+        ),
+    }
+}
+
+async fn check_rustdoc_json() -> DiagnosticResult {
+    // First check if rustdoc is available
+    match Command::new("rustdoc").arg("--version").output() {
+        Ok(output) if output.status.success() => {
+            let version = String::from_utf8_lossy(&output.stdout).trim().to_string();
+            
+            // Try to create a simple test for JSON generation
+            let temp_dir = match tempfile::tempdir() {
+                Ok(dir) => dir,
+                Err(_) => return DiagnosticResult::new(
+                    "Rustdoc JSON".to_string(),
+                    false,
+                    "Failed to create temporary directory for testing".to_string(),
+                    false,
+                ),
+            };
+            
+            let test_file = temp_dir.path().join("lib.rs");
+            if let Err(_) = fs::write(&test_file, "//! Test crate\npub fn test() {}") {
+                return DiagnosticResult::new(
+                    "Rustdoc JSON".to_string(),
+                    false,
+                    "Failed to create test file".to_string(),
+                    false,
+                );
+            }
+            
+            // Try to generate JSON documentation
+            match Command::new("rustdoc")
+                .args(&[
+                    "+nightly",
+                    "--output-format", "json",
+                    "--crate-name", "test",
+                    test_file.to_str().unwrap(),
+                ])
+                .output() {
+                Ok(json_output) if json_output.status.success() => {
+                    DiagnosticResult::new(
+                        "Rustdoc JSON".to_string(),
+                        true,
+                        format!("{} with JSON support", version),
+                        false,
+                    )
+                }
+                _ => DiagnosticResult::new(
+                    "Rustdoc JSON".to_string(),
+                    false,
+                    "JSON generation failed - ensure nightly toolchain is installed".to_string(),
+                    false,
+                ),
+            }
+        }
+        Ok(_) => DiagnosticResult::new(
+            "Rustdoc JSON".to_string(),
+            false,
+            "rustdoc command failed".to_string(),
+            false,
+        ),
+        Err(_) => DiagnosticResult::new(
+            "Rustdoc JSON".to_string(),
+            false,
+            "rustdoc not found in PATH".to_string(),
+            false,
+        ),
+    }
+}
+
+async fn check_git_installation() -> DiagnosticResult {
+    match Command::new("git").arg("--version").output() {
+        Ok(output) if output.status.success() => {
+            let version = String::from_utf8_lossy(&output.stdout).trim().to_string();
+            DiagnosticResult::new(
+                "Git".to_string(),
+                true,
+                version,
+                true,
+            )
+        }
+        Ok(_) => DiagnosticResult::new(
+            "Git".to_string(),
+            false,
+            "git command failed".to_string(),
+            true,
+        ),
+        Err(_) => DiagnosticResult::new(
+            "Git".to_string(),
+            false,
+            "git not found in PATH".to_string(),
+            true,
+        ),
+    }
+}
+
+async fn check_network_connectivity() -> DiagnosticResult {
+    let client = reqwest::Client::builder()
+        .timeout(std::time::Duration::from_secs(10))
+        .build()
+        .unwrap();
+    
+    // Test crates.io API
+    match client.get("https://crates.io/api/v1/crates/serde").send().await {
+        Ok(response) if response.status().is_success() => {
+            // Also test GitHub connectivity
+            match client.get("https://api.github.com").send().await {
+                Ok(gh_response) if gh_response.status().is_success() => {
+                    DiagnosticResult::new(
+                        "Network".to_string(),
+                        true,
+                        "crates.io and GitHub reachable".to_string(),
+                        false,
+                    )
+                }
+                _ => DiagnosticResult::new(
+                    "Network".to_string(),
+                    false,
+                    "crates.io reachable but GitHub unreachable".to_string(),
+                    false,
+                ),
+            }
+        }
+        _ => DiagnosticResult::new(
+            "Network".to_string(),
+            false,
+            "Unable to reach crates.io - check network connection".to_string(),
+            false,
+        ),
+    }
+}
+
+async fn check_cache_directory(cache_dir: Option<std::path::PathBuf>) -> DiagnosticResult {
+    let cache_path = match cache_dir {
+        Some(dir) => dir,
+        None => {
+            match dirs::home_dir() {
+                Some(home) => home.join(".rust-docs-mcp").join("cache"),
+                None => return DiagnosticResult::new(
+                    "Cache directory".to_string(),
+                    false,
+                    "Unable to determine home directory".to_string(),
+                    false,
+                ),
+            }
+        }
+    };
+    
+    // Check if directory exists or can be created
+    if !cache_path.exists() {
+        match fs::create_dir_all(&cache_path) {
+            Ok(_) => {},
+            Err(e) => return DiagnosticResult::new(
+                "Cache directory".to_string(),
+                false,
+                format!("Cannot create cache directory: {}", e),
+                false,
+            ),
+        }
+    }
+    
+    // Test write permissions
+    let test_file = cache_path.join(".test_write");
+    match fs::write(&test_file, "test") {
+        Ok(_) => {
+            let _ = fs::remove_file(&test_file);
+            
+            // Check available disk space
+            match fs::metadata(&cache_path) {
+                Ok(_) => {
+                    // We can't easily get disk space in a cross-platform way without additional dependencies
+                    // For now, just confirm it's writable
+                    DiagnosticResult::new(
+                        "Cache directory".to_string(),
+                        true,
+                        format!("{} (writable)", cache_path.display()),
+                        false,
+                    )
+                }
+                Err(_) => DiagnosticResult::new(
+                    "Cache directory".to_string(),
+                    true,
+                    format!("{} (writable)", cache_path.display()),
+                    false,
+                ),
+            }
+        }
+        Err(e) => DiagnosticResult::new(
+            "Cache directory".to_string(),
+            false,
+            format!("Directory not writable: {}", e),
+            false,
+        ),
+    }
+}
+
+async fn check_optional_dependencies() -> DiagnosticResult {
+    let mut messages = Vec::new();
+    
+    // Check for codesign on macOS
+    #[cfg(target_os = "macos")]
+    {
+        match Command::new("codesign").arg("--version").output() {
+            Ok(output) if output.status.success() => {
+                messages.push("codesign available".to_string());
+            }
+            _ => {
+                messages.push("codesign not available (optional for binary signing)".to_string());
+            }
+        }
+    }
+    
+    // If no optional dependencies to check, return success
+    if messages.is_empty() {
+        messages.push("No optional dependencies to check".to_string());
+    }
+    
+    DiagnosticResult::new(
+        "Optional dependencies".to_string(),
+        true,
+        messages.join(", "),
+        false,
+    )
+}
+
+pub fn print_results(results: &[DiagnosticResult]) {
+    println!("🔍 rust-docs-mcp doctor\n");
+    
+    let mut failed_count = 0;
+    let mut critical_failed = false;
+    
+    for result in results {
+        let icon = if result.success { "✅" } else { "❌" };
+        println!("{} {}: {}", icon, result.name, result.message);
+        
+        if !result.success {
+            failed_count += 1;
+            if result.critical {
+                critical_failed = true;
+            }
+        }
+    }
+    
+    if failed_count > 0 {
+        println!("\n[ERROR] Doctor found {} issue{}.", failed_count, if failed_count == 1 { "" } else { "s" });
+        
+        // Print specific error messages and suggestions
+        for result in results {
+            if !result.success {
+                match result.name.as_str() {
+                    "Rust toolchain" => {
+                        println!("\nRust toolchain is required. Please install Rust from https://rustup.rs/");
+                    }
+                    "Nightly toolchain" => {
+                        println!("\nNightly toolchain is required for rustdoc JSON generation. Install with:");
+                        println!("  rustup toolchain install nightly");
+                    }
+                    "Git" => {
+                        println!("\nGit is required for repository operations. Please install Git from https://git-scm.com/");
+                    }
+                    "Rustdoc JSON" => {
+                        println!("\nRustdoc JSON generation failed. Ensure nightly toolchain is properly installed:");
+                        println!("  rustup toolchain install nightly");
+                    }
+                    "Network" => {
+                        println!("\nNetwork connectivity issues detected. Check your internet connection.");
+                    }
+                    "Cache directory" => {
+                        println!("\nCache directory issues detected. Check file permissions and disk space.");
+                    }
+                    _ => {}
+                }
+            }
+        }
+        
+        println!("\nPlease fix the above errors before using rust-docs-mcp.");
+    } else {
+        println!("\n✅ All checks passed! rust-docs-mcp is ready to use.");
+    }
+}
+
+pub fn exit_code(results: &[DiagnosticResult]) -> i32 {
+    let mut has_failures = false;
+    let mut has_critical_failures = false;
+    
+    for result in results {
+        if !result.success {
+            has_failures = true;
+            if result.critical {
+                has_critical_failures = true;
+            }
+        }
+    }
+    
+    if has_critical_failures {
+        2 // Critical system dependency missing
+    } else if has_failures {
+        1 // One or more checks failed
+    } else {
+        0 // All checks passed
+    }
+}

--- a/rust-docs-mcp/src/lib.rs
+++ b/rust-docs-mcp/src/lib.rs
@@ -2,6 +2,7 @@ pub mod analysis;
 pub mod cache;
 pub mod deps;
 pub mod docs;
+pub mod rustdoc;
 pub mod service;
 
 pub use service::RustDocsService;

--- a/rust-docs-mcp/src/main.rs
+++ b/rust-docs-mcp/src/main.rs
@@ -173,17 +173,7 @@ async fn install_executable(target_dir: Option<PathBuf>, force: bool) -> Result<
     }
 
     // Run doctor command to verify the installation
-    println!("\n🔍 Running system diagnostics...\n");
-    let results = doctor::run_diagnostics(None).await?;
-    doctor::print_results(&results);
-    
-    let exit_code = doctor::exit_code(&results);
-    if exit_code != 0 {
-        println!("\n⚠️  Some diagnostic checks failed. Please address the issues above.");
-        println!("You can run 'rust-docs-mcp doctor' anytime to check system status.");
-    } else {
-        println!("\n✅ All system checks passed!");
-    }
+    doctor::run_and_print_diagnostics().await?;
 
     Ok(())
 }

--- a/rust-docs-mcp/src/main.rs
+++ b/rust-docs-mcp/src/main.rs
@@ -9,6 +9,7 @@ mod analysis;
 mod cache;
 mod deps;
 mod docs;
+mod doctor;
 mod service;
 mod update;
 use service::RustDocsService;
@@ -48,6 +49,8 @@ enum Commands {
         #[arg(long)]
         branch: Option<String>,
     },
+    /// Verify system environment and dependencies
+    Doctor,
 }
 
 #[tokio::main]
@@ -57,7 +60,7 @@ async fn main() -> Result<()> {
 
     // Handle subcommands
     if let Some(command) = args.command {
-        return handle_command(command).await;
+        return handle_command(command, args.cache_dir).await;
     }
 
     // Initialize tracing to stderr to avoid conflicts with stdio transport
@@ -85,7 +88,7 @@ async fn main() -> Result<()> {
     Ok(())
 }
 
-async fn handle_command(command: Commands) -> Result<()> {
+async fn handle_command(command: Commands, cache_dir: Option<PathBuf>) -> Result<()> {
     match command {
         Commands::Install { target_dir, force } => install_executable(target_dir, force).await,
         Commands::Update {
@@ -93,6 +96,7 @@ async fn handle_command(command: Commands) -> Result<()> {
             repo_url,
             branch,
         } => update::update_executable(target_dir, repo_url, branch).await,
+        Commands::Doctor => handle_doctor_command(cache_dir).await,
     }
 }
 
@@ -163,4 +167,10 @@ async fn install_executable(target_dir: Option<PathBuf>, force: bool) -> Result<
     }
 
     Ok(())
+}
+
+async fn handle_doctor_command(cache_dir: Option<PathBuf>) -> Result<()> {
+    let results = doctor::run_diagnostics(cache_dir).await?;
+    doctor::print_results(&results);
+    process::exit(doctor::exit_code(&results));
 }

--- a/rust-docs-mcp/src/main.rs
+++ b/rust-docs-mcp/src/main.rs
@@ -10,6 +10,7 @@ mod cache;
 mod deps;
 mod docs;
 mod doctor;
+mod rustdoc;
 mod service;
 mod update;
 use service::RustDocsService;

--- a/rust-docs-mcp/src/main.rs
+++ b/rust-docs-mcp/src/main.rs
@@ -172,6 +172,19 @@ async fn install_executable(target_dir: Option<PathBuf>, force: bool) -> Result<
         }
     }
 
+    // Run doctor command to verify the installation
+    println!("\n🔍 Running system diagnostics...\n");
+    let results = doctor::run_diagnostics(None).await?;
+    doctor::print_results(&results);
+    
+    let exit_code = doctor::exit_code(&results);
+    if exit_code != 0 {
+        println!("\n⚠️  Some diagnostic checks failed. Please address the issues above.");
+        println!("You can run 'rust-docs-mcp doctor' anytime to check system status.");
+    } else {
+        println!("\n✅ All system checks passed!");
+    }
+
     Ok(())
 }
 

--- a/rust-docs-mcp/src/main.rs
+++ b/rust-docs-mcp/src/main.rs
@@ -152,7 +152,8 @@ async fn install_executable(target_dir: Option<PathBuf>, force: bool) -> Result<
 
     // Check if target directory is in PATH
     if let Ok(path_var) = env::var("PATH") {
-        let paths: Vec<&str> = path_var.split(':').collect();
+        let path_separator = if cfg!(windows) { ';' } else { ':' };
+        let paths: Vec<&str> = path_var.split(path_separator).collect();
         let target_dir_str = target_dir.to_string_lossy();
 
         if !paths.iter().any(|&p| p == target_dir_str) {

--- a/rust-docs-mcp/src/main.rs
+++ b/rust-docs-mcp/src/main.rs
@@ -51,7 +51,11 @@ enum Commands {
         branch: Option<String>,
     },
     /// Verify system environment and dependencies
-    Doctor,
+    Doctor {
+        /// Output results in JSON format for programmatic consumption
+        #[arg(long)]
+        json: bool,
+    },
 }
 
 #[tokio::main]
@@ -97,7 +101,7 @@ async fn handle_command(command: Commands, cache_dir: Option<PathBuf>) -> Result
             repo_url,
             branch,
         } => update::update_executable(target_dir, repo_url, branch).await,
-        Commands::Doctor => handle_doctor_command(cache_dir).await,
+        Commands::Doctor { json } => handle_doctor_command(cache_dir, json).await,
     }
 }
 
@@ -171,8 +175,14 @@ async fn install_executable(target_dir: Option<PathBuf>, force: bool) -> Result<
     Ok(())
 }
 
-async fn handle_doctor_command(cache_dir: Option<PathBuf>) -> Result<()> {
+async fn handle_doctor_command(cache_dir: Option<PathBuf>, json_output: bool) -> Result<()> {
     let results = doctor::run_diagnostics(cache_dir).await?;
-    doctor::print_results(&results);
+    
+    if json_output {
+        doctor::print_results_json(&results)?;
+    } else {
+        doctor::print_results(&results);
+    }
+    
     process::exit(doctor::exit_code(&results));
 }

--- a/rust-docs-mcp/src/rustdoc.rs
+++ b/rust-docs-mcp/src/rustdoc.rs
@@ -88,23 +88,39 @@ pub async fn get_rustdoc_version() -> Result<String> {
     Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
 }
 
-/// Run cargo rustdoc with JSON output for a crate
-pub async fn run_cargo_rustdoc_json(source_path: &Path) -> Result<()> {
+/// Run cargo rustdoc with JSON output for a crate or specific package
+pub async fn run_cargo_rustdoc_json(source_path: &Path, package: Option<&str>) -> Result<()> {
     validate_toolchain().await?;
     
-    tracing::debug!("Running cargo rustdoc with JSON output in {}", source_path.display());
+    let log_msg = match package {
+        Some(pkg) => format!("Running cargo rustdoc with JSON output for package {} in {}", pkg, source_path.display()),
+        None => format!("Running cargo rustdoc with JSON output in {}", source_path.display()),
+    };
+    tracing::debug!("{}", log_msg);
+    
+    let mut args = vec![
+        format!("+{}", REQUIRED_TOOLCHAIN),
+        "rustdoc".to_string(),
+    ];
+    
+    // Add package-specific arguments if provided
+    if let Some(pkg) = package {
+        args.push("-p".to_string());
+        args.push(pkg.to_string());
+    }
+    
+    // Add remaining arguments
+    args.extend_from_slice(&[
+        "--all-features".to_string(),
+        "--".to_string(),
+        "--output-format".to_string(),
+        "json".to_string(),
+        "-Z".to_string(),
+        "unstable-options".to_string(),
+    ]);
     
     let output = Command::new("cargo")
-        .args([
-            &format!("+{}", REQUIRED_TOOLCHAIN),
-            "rustdoc",
-            "--all-features",
-            "--",
-            "--output-format",
-            "json",
-            "-Z",
-            "unstable-options",
-        ])
+        .args(&args)
         .current_dir(source_path)
         .output()
         .context("Failed to run cargo rustdoc")?;
@@ -117,40 +133,6 @@ pub async fn run_cargo_rustdoc_json(source_path: &Path) -> Result<()> {
     Ok(())
 }
 
-/// Run cargo rustdoc with JSON output for a specific package
-pub async fn run_cargo_rustdoc_json_for_package(source_path: &Path, package_name: &str) -> Result<()> {
-    validate_toolchain().await?;
-    
-    tracing::debug!(
-        "Running cargo rustdoc with JSON output for package {} in {}", 
-        package_name, 
-        source_path.display()
-    );
-    
-    let output = Command::new("cargo")
-        .args([
-            &format!("+{}", REQUIRED_TOOLCHAIN),
-            "rustdoc",
-            "-p",
-            package_name,
-            "--all-features",
-            "--",
-            "--output-format",
-            "json",
-            "-Z",
-            "unstable-options",
-        ])
-        .current_dir(source_path)
-        .output()
-        .context("Failed to run cargo rustdoc")?;
-
-    if !output.status.success() {
-        let stderr = String::from_utf8_lossy(&output.stderr);
-        bail!("Failed to generate documentation: {}", stderr);
-    }
-
-    Ok(())
-}
 
 #[cfg(test)]
 mod tests {

--- a/rust-docs-mcp/src/rustdoc.rs
+++ b/rust-docs-mcp/src/rustdoc.rs
@@ -1,0 +1,176 @@
+//! Unified rustdoc JSON generation functionality
+//!
+//! Provides consistent rustdoc JSON generation across the application,
+//! including toolchain validation and command execution.
+
+use anyhow::{Context, Result, bail};
+use std::path::Path;
+use std::process::Command;
+
+/// The pinned nightly toolchain version compatible with rustdoc-types 0.53.0
+pub const REQUIRED_TOOLCHAIN: &str = "nightly-2025-06-23";
+
+/// Check if the required nightly toolchain is available
+pub async fn validate_toolchain() -> Result<()> {
+    let output = Command::new("rustup")
+        .args(["toolchain", "list"])
+        .output()
+        .context("Failed to run rustup toolchain list")?;
+
+    if !output.status.success() {
+        bail!("Failed to check available toolchains");
+    }
+
+    let toolchains = String::from_utf8_lossy(&output.stdout);
+    if !toolchains.contains(REQUIRED_TOOLCHAIN) {
+        bail!(
+            "Required toolchain {} is not installed. Please run: rustup toolchain install {}",
+            REQUIRED_TOOLCHAIN,
+            REQUIRED_TOOLCHAIN
+        );
+    }
+
+    tracing::debug!("Validated toolchain {} is available", REQUIRED_TOOLCHAIN);
+    Ok(())
+}
+
+/// Test rustdoc JSON functionality with a simple test file
+pub async fn test_rustdoc_json() -> Result<()> {
+    // First validate the toolchain
+    validate_toolchain().await?;
+    
+    // Create a temporary directory and test file
+    let temp_dir = tempfile::tempdir()
+        .context("Failed to create temporary directory for testing")?;
+    
+    let test_file = temp_dir.path().join("lib.rs");
+    std::fs::write(&test_file, "//! Test crate\npub fn test() {}")
+        .context("Failed to create test file")?;
+    
+    let test_file_str = test_file
+        .to_str()
+        .ok_or_else(|| anyhow::anyhow!("Test file path contains invalid UTF-8"))?;
+    
+    tracing::debug!("Testing rustdoc JSON generation with {}", REQUIRED_TOOLCHAIN);
+    
+    // Try to generate JSON documentation using the pinned toolchain
+    let output = Command::new("rustdoc")
+        .args(&[
+            &format!("+{}", REQUIRED_TOOLCHAIN),
+            "-Z", "unstable-options",
+            "--output-format", "json",
+            "--crate-name", "test",
+            test_file_str,
+        ])
+        .output()
+        .context("Failed to run rustdoc")?;
+    
+    if !output.status.success() {
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        bail!("JSON generation failed: {}", stderr);
+    }
+    
+    tracing::debug!("Successfully tested rustdoc JSON generation");
+    Ok(())
+}
+
+/// Get rustdoc version information
+pub async fn get_rustdoc_version() -> Result<String> {
+    let output = Command::new("rustdoc")
+        .arg("--version")
+        .output()
+        .context("Failed to run rustdoc --version")?;
+
+    if !output.status.success() {
+        bail!("rustdoc command failed");
+    }
+
+    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
+}
+
+/// Run cargo rustdoc with JSON output for a crate
+pub async fn run_cargo_rustdoc_json(source_path: &Path) -> Result<()> {
+    validate_toolchain().await?;
+    
+    tracing::debug!("Running cargo rustdoc with JSON output in {}", source_path.display());
+    
+    let output = Command::new("cargo")
+        .args([
+            &format!("+{}", REQUIRED_TOOLCHAIN),
+            "rustdoc",
+            "--all-features",
+            "--",
+            "--output-format",
+            "json",
+            "-Z",
+            "unstable-options",
+        ])
+        .current_dir(source_path)
+        .output()
+        .context("Failed to run cargo rustdoc")?;
+
+    if !output.status.success() {
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        bail!("Failed to generate documentation: {}", stderr);
+    }
+
+    Ok(())
+}
+
+/// Run cargo rustdoc with JSON output for a specific package
+pub async fn run_cargo_rustdoc_json_for_package(source_path: &Path, package_name: &str) -> Result<()> {
+    validate_toolchain().await?;
+    
+    tracing::debug!(
+        "Running cargo rustdoc with JSON output for package {} in {}", 
+        package_name, 
+        source_path.display()
+    );
+    
+    let output = Command::new("cargo")
+        .args([
+            &format!("+{}", REQUIRED_TOOLCHAIN),
+            "rustdoc",
+            "-p",
+            package_name,
+            "--all-features",
+            "--",
+            "--output-format",
+            "json",
+            "-Z",
+            "unstable-options",
+        ])
+        .current_dir(source_path)
+        .output()
+        .context("Failed to run cargo rustdoc")?;
+
+    if !output.status.success() {
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        bail!("Failed to generate documentation: {}", stderr);
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_get_rustdoc_version() {
+        // This test will pass if rustdoc is installed
+        let result = get_rustdoc_version().await;
+        // We can't guarantee the success state in all environments
+        // but we can verify it returns a valid result
+        assert!(result.is_ok() || result.is_err());
+    }
+
+    #[tokio::test]
+    async fn test_validate_toolchain() {
+        // This test will pass if rustup is installed
+        let result = validate_toolchain().await;
+        // We can't guarantee the toolchain is installed in all environments
+        // but we can verify it returns a valid result
+        assert!(result.is_ok() || result.is_err());
+    }
+}

--- a/rust-docs-mcp/src/update.rs
+++ b/rust-docs-mcp/src/update.rs
@@ -6,6 +6,7 @@
 use anyhow::{Context, Result};
 use std::path::PathBuf;
 use std::process::Command;
+use crate::doctor;
 
 /// Update rust-docs-mcp to the latest version from GitHub
 pub async fn update_executable(
@@ -103,7 +104,21 @@ pub async fn update_executable(
     println!("  rust-docs-mcp                # Start MCP server");
     println!("  rust-docs-mcp install        # Install/update to PATH");
     println!("  rust-docs-mcp update         # Update to latest version");
+    println!("  rust-docs-mcp doctor         # Verify system environment");
     println!("  rust-docs-mcp --help         # Show help");
+
+    // Run doctor command to verify the update
+    println!("\n🔍 Running system diagnostics...\n");
+    let results = doctor::run_diagnostics(None).await?;
+    doctor::print_results(&results);
+    
+    let exit_code = doctor::exit_code(&results);
+    if exit_code != 0 {
+        println!("\n⚠️  Some diagnostic checks failed. Please address the issues above.");
+        println!("You can run 'rust-docs-mcp doctor' anytime to check system status.");
+    } else {
+        println!("\n✅ All system checks passed!");
+    }
 
     Ok(())
 }
@@ -198,7 +213,8 @@ fn check_path_and_advise(target_dir: &PathBuf) -> Result<()> {
     use std::env;
 
     if let Ok(path_var) = env::var("PATH") {
-        let paths: Vec<&str> = path_var.split(':').collect();
+        let path_separator = if cfg!(windows) { ';' } else { ':' };
+        let paths: Vec<&str> = path_var.split(path_separator).collect();
         let target_dir_str = target_dir.to_string_lossy();
 
         if !paths.iter().any(|&p| p == target_dir_str) {

--- a/rust-docs-mcp/src/update.rs
+++ b/rust-docs-mcp/src/update.rs
@@ -108,17 +108,7 @@ pub async fn update_executable(
     println!("  rust-docs-mcp --help         # Show help");
 
     // Run doctor command to verify the update
-    println!("\n🔍 Running system diagnostics...\n");
-    let results = doctor::run_diagnostics(None).await?;
-    doctor::print_results(&results);
-    
-    let exit_code = doctor::exit_code(&results);
-    if exit_code != 0 {
-        println!("\n⚠️  Some diagnostic checks failed. Please address the issues above.");
-        println!("You can run 'rust-docs-mcp doctor' anytime to check system status.");
-    } else {
-        println!("\n✅ All system checks passed!");
-    }
+    doctor::run_and_print_diagnostics().await?;
 
     Ok(())
 }

